spec/defines/rpm_gpg_key_spec.rb

```diff
@@ -1,119 +1,73 @@
-        require:   "File[#{params[:path]}]",
-        logoutput: 'on_failure'
-      )
-    end
-  end
-
-  context 'operatingsystemmajrelease => 7' do
-    let :facts do
-      default_facts.merge(
-        operatingsystemrelease:    '7.0.1406',
-        operatingsystemmajrelease: '7'
-      )
-    end
-
-    let :title do
-      'EPEL-7'
-    end
-
-    let :params do
+  test_on = {
+    supported_os: [
       {
-        path: '/etc/pki/rpm-gpg/RPM-GPG-KEY-EPEL-7'
+        'operatingsystem'        => 'RedHat',
+        'operatingsystemrelease' => %w[5 6 7 8]
       }
-    end
-
-    it do
-      is_expected.to contain_exec("import-#{title}").with(
-        path:      '/bin:/usr/bin:/sbin:/usr/sbin',
-        command:   "rpm --import #{params[:path]}",
-        unless:    "rpm -q gpg-pubkey-$(echo $(gpg -q --throw-keyids --keyid-format short < #{params[:path]}) | grep pub | cut -f2 -d/ | cut -f1 -d' ' | tr '[A-Z]' '[a-z]')",
-        require:   "File[#{params[:path]}]",
-        logoutput: 'on_failure'
-      )
-    end
-  end
-
-  context 'operatingsystemmajrelease => 6' do
-    let :facts do
-      default_facts.merge(
-        operatingsystemrelease:    '6.4',
-        operatingsystemmajrelease: '6'
-      )
-    end
-
-    let :title do
-      'EPEL-6'
-    end
-
-    let :params do
-      {
-        path: '/etc/pki/rpm-gpg/RPM-GPG-KEY-EPEL-6'
-      }
-    end
-
-    it do
-      is_expected.to contain_exec("import-#{title}").with(
-        path:      '/bin:/usr/bin:/sbin:/usr/sbin',
-        command:   "rpm --import #{params[:path]}",
-        unless:    "rpm -q gpg-pubkey-$(echo $(gpg -q --throw-keyids --keyid-format short < #{params[:path]}) | grep pub | cut -f2 -d/ | cut -f1 -d' ' | tr '[A-Z]' '[a-z]')",
-        require:   "File[#{params[:path]}]",
-        logoutput: 'on_failure'
-      )
-    end
-  end
-
-  context 'operatingsystemmajrelease => 5' do
-    let :facts do
-      default_facts.merge(
-        operatingsystemrelease:    '5.9',
-        operatingsystemmajrelease: '5'
-      )
-    end
-
-    let :title do
-      'EPEL-5'
-    end
-
-    let :params do
-      {
-        path: '/etc/pki/rpm-gpg/RPM-GPG-KEY-EPEL-5'
-      }
-    end
-
-    it do
-      is_expected.to contain_exec("import-#{title}").with(
-        path:      '/bin:/usr/bin:/sbin:/usr/sbin',
-        command:   "rpm --import #{params[:path]}",
-        unless:    "rpm -q gpg-pubkey-$(echo $(gpg -q --throw-keyids --keyid-format short < #{params[:path]}) | grep pub | cut -f2 -d/ | cut -f1 -d' ' | tr '[A-Z]' '[a-z]')",
-        require:   "File[#{params[:path]}]",
-        logoutput: 'on_failure'
-      )
+    ]
+  }
+  on_supported_os(test_on).each do |os, os_facts|
+    context "on #{os}" do
+      let(:facts) { os_facts }
+
+      case os_facts[:operatingsystemmajrelease]
+      when '5'
+        let :title do
+          'EPEL-5'
+        end
+        let :pre_condition do
+          'file { "/etc/pki/rpm-gpg/RPM-GPG-KEY-EPEL-5": }'
```

Comment:
Puppet 5 complains if you reference a resource not in your catalog, so I had to mock it.